tests/components/teslemetry/test_services.py

```diff
@@ -0,0 +1,302 @@
+            blocking=True,
+        )
+        speed_limit_activate.assert_called_once()
+
+    with patch(
+        "homeassistant.components.teslemetry.VehicleSpecific.speed_limit_deactivate",
+        return_value=COMMAND_OK,
+    ) as speed_limit_deactivate:
+        await hass.services.async_call(
+            DOMAIN,
+            SERVICE_SPEED_LIMIT,
+            {
+                CONF_DEVICE_ID: vehicle_device,
+                ATTR_ENABLE: False,
+                ATTR_PIN: 1234,
+            },
+            blocking=True,
+        )
+        speed_limit_deactivate.assert_called_once()
+
+    with patch(
+        "homeassistant.components.teslemetry.EnergySpecific.time_of_use_settings",
+        return_value=COMMAND_OK,
+    ) as set_time_of_use:
+        await hass.services.async_call(
+            DOMAIN,
+            SERVICE_TIME_OF_USE,
+            {
+                CONF_DEVICE_ID: energy_device,
+                ATTR_TOU_SETTINGS: {},
+            },
+            blocking=True,
+        )
+        set_time_of_use.assert_called_once()
+
+    with (
+        patch(
+            "homeassistant.components.teslemetry.EnergySpecific.time_of_use_settings",
+            return_value=COMMAND_ERROR,
+        ) as set_time_of_use,
+        pytest.raises(HomeAssistantError),
+    ):
+        await hass.services.async_call(
+            DOMAIN,
+            SERVICE_TIME_OF_USE,
+            {
+                CONF_DEVICE_ID: energy_device,
+                ATTR_TOU_SETTINGS: {},
+            },
+            blocking=True,
+        )
+
+
+async def test_service_validation_errors(
+    hass: HomeAssistant,
+) -> None:
+    """Tests that the custom services handle bad data."""
+
+    entry = await setup_platform(hass)
+    entity_registry = er.async_get(hass)
+
+    # Get a vehicle device ID
+    vehicle_device = entity_registry.async_get("sensor.test_charging").device_id
+    energy_device = entity_registry.async_get(
+        "sensor.energy_site_battery_power"
+    ).device_id
+
+    # Bad device ID
+    with pytest.raises(ServiceValidationError):
+        await hass.services.async_call(
+            DOMAIN,
+            SERVICE_NAVIGATE_ATTR_GPS_REQUEST,
+            {
+                CONF_DEVICE_ID: "nope",
+                ATTR_GPS: {CONF_LATITUDE: lat, CONF_LONGITUDE: lon},
+            },
+            blocking=True,
+        )
+
+    # Device with no config entry
+    with (
+        pytest.raises(ServiceValidationError),
+        patch(
+            "homeassistant.components.teslemetry.services.async_get_device_for_service_call",
+            return_value=DeviceEntry(),
+        ),
+    ):
+        await hass.services.async_call(
+            DOMAIN,
+            SERVICE_NAVIGATE_ATTR_GPS_REQUEST,
+            {
+                CONF_DEVICE_ID: vehicle_device,
+                ATTR_GPS: {CONF_LATITUDE: lat, CONF_LONGITUDE: lon},
+            },
+            blocking=True,
+        )
+
+    entry.runtime_data = TeslemetryData([], [], [])
```

Comment:
This seems very strange to me?
We typically don't allow to mock internals and also can this even happen as I would think the integrations would not even start then?

Maybe you're intending to do tests for the services if the config entry is not there at all or unloaded?

Reply:
The helper function that finds the specific part of the runtime data for a device id has failure cases if it can't find the specific VIN. I can avoid this type of test if I also remove that safety from the code.

I think this specific test has found the device, found the config data, but cannot find the VIN in the config data.

You're right that it should be impossible. Should I remove the failure case from the code so that it doesn't need to be tested?